docs/architecture.md

```diff
@@ -0,0 +1,228 @@
+# Loki's Architecture
+
+This document will expand on the information detailed in the [Loki
+Overview](overview/README.md).
+
+## Multi Tenancy
+
+All data both in memory and in long-term storage is partitioned by a tenant ID,
+pulled from the `X-Scope-OrgID` header from the request when Loki is run in
+multi-tenant mode. When Loki is **not** in multi-tenant mode, the header
+is ignored and the tenant ID is set to "fake", which will appear in the index
+and in stored chunks.
+
+## Modes of Operation
+
+![modes_diagram](modes_of_operation.png)
+
+Loki has a set of components (defined below in [Components](#components)) which
+are internally referred to as modules. Each component spawns a gRPC server for
+internal traffic and an HTTP/1 server for external API requests. All components
+come with an HTTP/1 server, but most only expose readiness and health endpoints.
+
+While each component can be run in a separate process, Loki also supports running
+all components in a single process. Running all components in a single process is
+referred to as "single process" mode (or sometimes "single binary" or monolithic
+mode), while the other mode is the "horizontally" scalable, or microservices
+mode.
+
+When Loki runs in single process mode, individual components continue to
+communicate to one another over gRPC using the gRPC listen port of the overall
+process.
+
+Single process mode is ideally suited for local development, small workloads,
+and for evaluation purposes. Single-process mode can be scaled with multiple
```

Comment:
How do you feel about referring to it as monolithic mode here? In this context, I'm trying to explain that while you can run multiple instances of Loki with all modules enabled, they can't utilize local storage and you can't configure it to run more of a specific module.

Reply:
I personally consider "monolithic vs microservices" an implementation detail. From my perspective, the real difference is the operational mode is "single node" vs "cluster":

- Single node: all Loki services run - as a single process / binary - on a single node
- Cluster: distribute Loki services across multiple nodes and - optionally - run multiple replicas of a single service in order to achieve HA and/or scalability (via sharding)

When running in the cluster mode, the limitations are:
- You can't use the local storage for indexes and chunks

Is this what you meant to describe or you're describing something different?

Reply:
I'm describing something just slightly different. "monolithic" mode is just when you pass `-target=all` (or have `target: all` set in the config file), but it still supports clustering. 

Microservices mode is when you have _at least_ 3 instances of Loki, one with `-target=querier`, one with `-target=ingester`, and one with `-target=distributor`.

People deploying with the Helm charts are running a Loki with `-target=all`, and we have seen them try to scale up the number of replicas. This is supported, but: 

- You can't use the local storage for the index and chunks 
- You can't _just_ scale up the number of a single component, since the components are coupled together in each replica.

Does that make sense?

Reply:
It definitely makes sense and now I do understand what you mean. You may try to improve the doc of this part, being a bit more verbose as you've been (patiently) explaining it to me 👍